File: Leetcode/1123 - Lowest Common Ancestor of Deepest Leaves/solution.py
class Solution:
    def lcaDeepestLeaves(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        return self.getLcaDeepestLeaves(root, 1)[0]
    
    def getLcaDeepestLeaves(self, root: Optional[TreeNode], currentDepth: int) -> Tuple[Optional[TreeNode], int]:
        if not root.left and not root.right:
            return root, currentDepth
        else:
            leftLCA = None
            rightLCA = None
            leftDepth = 0
            rightDepth = 0

            if root.left:
                leftLCA, leftDepth = self.getLcaDeepestLeaves(root.left, currentDepth + 1)

            if root.right:
                rightLCA, rightDepth = self.getLcaDeepestLeaves(root.right, currentDepth + 1)

            if leftLCA and rightLCA:
                if leftDepth > rightDepth:
                    return leftLCA, leftDepth
                elif rightDepth > leftDepth:
                    return rightLCA, rightDepth
                else:
                    return root, leftDepth
            elif leftLCA:
                return leftLCA, leftDepth
            else:
                return rightLCA, rightDepth

File: Leetcode/1123 - Lowest Common Ancestor of Deepest Leaves/solution.js
var lcaDeepestLeaves = function (root) {
    return getLcaDeepestLeaves(root, 1)[0];
};

var getLcaDeepestLeaves = function (root, currentDepth) {
    if (!(root.left || root.right)) {
        return [root, currentDepth];
    } else {
        let leftLCA = null;
        let rightLCA = null;
        let leftDepth = 0;
        let rightDepth = 0;

        if (root.left) {
            const res = getLcaDeepestLeaves(root.left, currentDepth + 1);
            leftLCA = res[0];
            leftDepth = res[1];
        }
        if (root.right) {
            const res = getLcaDeepestLeaves(root.right, currentDepth + 1);
            rightLCA = res[0];
            rightDepth = res[1];
        }

        if (leftLCA && rightLCA) {
            if (leftDepth > rightDepth) {
                return [leftLCA, leftDepth];
            } else if (rightDepth > leftDepth) {
                return [rightLCA, rightDepth];
            } else {
                return [root, leftDepth];
            }
        } else if (leftLCA) {
            return [leftLCA, leftDepth];
        } else {
            return [rightLCA, rightDepth];
        }
    }
}